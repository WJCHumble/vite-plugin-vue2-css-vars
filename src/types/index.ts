export interface VueQuery {
  vue: boolean
  src: boolean
  index: number
  type: string
}

export interface PaserVueRequestResult {
  filename: string
  query: VueQuery
}

export interface CssVarsPluginOptions {
  id: string
  isProd?: boolean
}
