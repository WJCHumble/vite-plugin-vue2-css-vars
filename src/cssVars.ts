

import { CssVarsPluginOptions } from './types'
import { PluginCreator } from 'postcss'

const cssVarRE = /\bv-bind\(\s*(?:'([^']+)'|"([^"]+)"|([^'"][^)]*))\s*\)/g

function genVarName(id: string, raw: string, isProd?: boolean): string {
  return `${id}-${raw.replace(/([^\w-])/g, '_')}`
}

export const cssVarsPlugin: PluginCreator<CssVarsPluginOptions> = (opts) => {
  const { id } = opts
  return {
    postcssPlugin: 'vue-sfc-vars',
    Declaration(decl) {
      // rewrite CSS variables
      if (cssVarRE.test(decl.value)) {
        decl.value = decl.value.replace(cssVarRE, (_, $1, $2, $3) => {
          return `var(--${genVarName(id, $1 || $2 || $3)})`
        })
      }
    },
  }
}
cssVarsPlugin.postcss = true

export function parseCssVars(styles: Array<any>): string[] {
  const vars: string[] = []
  styles.forEach((style) => {
    let match
    // eslint-disable-next-line no-cond-assign
    while ((match = cssVarRE.exec(style.content)))
      vars.push(match[1] || match[2] || match[3])
  })
  return vars
}

export function genCssVarsFromList(
  vars: string[],
  id: string,
  isProd: boolean
): string {
  return `{\n  ${vars
    .map((key) => `"${genVarName(id, key, isProd)}": (${key})`)
    .join(',\n  ')}\n}`
}

export function genCssVarsCode(
  attrs: string[],
  id: string
): string {
  const watchCode = attrs.map(attr => `
      this.$watch('${attr}', function(nVal) { 
        const vnode = this._vnode 
        this._setVnodeVar(vnode, '--${id}-${attr}', nVal)
      }, { immediate: true })\n`)

    return `
  mixins: [{
    mounted: function () {
      if (!this._setVnodeVar) {
        this._setVnodeVar = function(vnode, cssVar, value) {
          if (!vnode.tag) {
            return
          }

          vnode.elm.style.setProperty(cssVar, value)\n
          if (vnode.children && vnode.children.length) {
            vnode.children.forEach((childVnode) => {
              this._setVnodeVar(childVnode, cssVar, value);
            });
          }
        }
      }
      ${watchCode}
    }
  }],`
}