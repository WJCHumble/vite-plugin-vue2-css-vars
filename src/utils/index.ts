import qs from 'querystring'
import { PluginCreator } from 'postcss'
import { PaserVueRequestResult, VueQuery, CssVarsPluginOptions } from '../types'

const cssVarRE = /\bv-bind\(\s*(?:'([^']+)'|"([^"]+)"|([^'"][^)]*))\s*\)/g

export function parseVueRequest(id): PaserVueRequestResult {
  const [filename, rawQuery] = id.split('?', 2)
  const query: VueQuery = qs.parse(rawQuery) as any
  if (query.vue != null)
    query.vue = true

  if (query.src != null)
    query.src = true

  if (query.index != null)
    query.index = Number(query.index)

  return {
    filename,
    query,
  }
}

function genVarName(id: string, raw: string, isProd?: boolean): string {
  return `${id}-${raw.replace(/([^\w-])/g, '_')}`
}

export const cssVarsPlugin: PluginCreator<CssVarsPluginOptions> = (opts) => {
  const { id } = opts
  return {
    postcssPlugin: 'vue-sfc-vars',
    Declaration(decl) {
      // rewrite CSS variables
      if (cssVarRE.test(decl.value)) {
        decl.value = decl.value.replace(cssVarRE, (_, $1, $2, $3) => {
          return `var(--${genVarName(id, $1 || $2 || $3)})`
        })
      }
    },
  }
}
cssVarsPlugin.postcss = true
