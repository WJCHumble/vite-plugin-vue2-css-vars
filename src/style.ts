
import { RawSourceMap } from '@vue/component-compiler-utils/dist/types';
import postcss from 'postcss'
import { TransformResult } from 'vite';
import { cssVarsPlugin } from "./cssVars"

export async function transformStyle(
  code: string, 
  id: string
): Promise<TransformResult> {
  const { css } = await postcss([cssVarsPlugin({ id })]).process(code, {from: undefined});
  return {
    code: css,
    map: null
  };
}