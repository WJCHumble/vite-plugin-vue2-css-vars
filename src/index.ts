import { Plugin } from 'vite'
import { createFilter } from '@rollup/pluginutils'
import compiler from 'vue-template-compiler'
import { parse } from '@vue/component-compiler-utils'
import { VueTemplateCompiler } from '@vue/component-compiler-utils/dist/types'
import postcss from 'postcss'
import hashsum from 'hash-sum'
import { parseVueRequest, cssVarsPlugin, parseCssVars } from './utils'

export default function vitePluginVue2CssVars(Options: Record<string, string> = {}): Plugin {
  const filter = createFilter(
    Options.include || /\.vue$/,
    Options.exclude,
  )

  return {
    name: 'vite-plugin-vue2-css-vars',
    async transform(code: string, id: string) {
      const { query, filename } = parseVueRequest(id)
      const shortId = hashsum(filename)
      // resolve style
      if (query.type && query.type === "style") {
        const { css } = await postcss([cssVarsPlugin({ id: shortId })]).process(code);
        return {
          code: css,
          map: null
        };
      }
      // resolve the .vue file which is without query
      if (!filter(filename) || Object.keys(query).length) {
        return
      }

      const { template, styles, script } = parse({
        source: code,
        filename: filename,
        needMap: true,
        compiler: compiler as VueTemplateCompiler
      });

      const { css } = await postcss([cssVarsPlugin({ id: shortId })]).process(styles[0].content)
      const attrs = parseCssVars(styles)

      // inject vueCssVars code
      const watchVarCode = attrs.map(attr => `
    this.$watch('${attr}', function(nVal) { 
      const vnode = this._vnode 
      this._setVnodeVar(vnode, '--${shortId}-${attr}', nVal)
    }, { immediate: true })`).join(";")
        const insertIndex = script.content.indexOf("{") + 1
        const useCssVarsCode = `
  mounted() {
    if (!this._setVnodeVar) {
      this._setVnodeVar = function(vnode, cssVar, value) {
        if (!vnode.tag) {
          return
        }

        vnode.elm.style.setProperty(cssVar, value)\n
        if (vnode.children && vnode.children.length) {
          vnode.children.forEach((childVnode) => {
            this._setVnodeVar(childVnode, cssVar, value);
          });
        }
      }
    }
    ${watchVarCode}
  },`
      // rewrite sfc file
      code = `<template>
${template.content}
</template>
<script>
${script.content.slice(0, insertIndex)}${useCssVarsCode}${script.content.slice(insertIndex)}
</script>
<style scoped ${Object.keys(styles[0].attrs).filter(key => key !== 'scoped')
                                            .map(key => `${key}='${styles[0].attrs[key]}'`)
                                            .join(" ")}>
${css}
</style>`

      return {
        code: code,
        map: null
      }
    },
  }
}
