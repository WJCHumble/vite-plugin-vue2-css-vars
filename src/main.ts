import { VueTemplateCompiler } from '@vue/component-compiler-utils/dist/types'
import { parse as SFCParse } from '@vue/component-compiler-utils'
import * as vueTemplateCompiler from 'vue-template-compiler'
import { transformStyle } from "./style"
import { parseCssVars, genCssVarsCode } from "./cssVars"
import { TransformResult } from 'vite'
import { init, parse } from "es-module-lexer"

export async function transformMain(
  code: string, 
  filename: string, 
  id: string,
  isProd: boolean
 ): Promise<TransformResult> {
  const { template, styles, script } = SFCParse({
    source: code,
    filename: filename,
    needMap: true,
    compiler: vueTemplateCompiler as VueTemplateCompiler
  });

  // sfc hasn't style block
  if (!styles.length) {
    return
  }

  const {code: css} = await transformStyle(styles[0].content, id)
  const attrs = await parseCssVars(styles)
  // In prod mode, will not auto inject useCssVars code without vars attr in <style> block
  if (isProd && !styles[0].attrs.vars) {
    return
  }
  // use v-bind(), but no use vars attr in <style> block
  if (!styles[0].attrs.vars && attrs.length) {
    throw new Error(`If you use v-bind() in <style> block, you should add vars attr in <style>.`)
  }
  // resolve the import
  await init
  const [imports] = parse(script.content)

  let importCode = "", exportDefaultCode = script.content
  if (imports.length) {
    importCode = script.content.substring(imports[0]["ss"], imports[imports.length - 1]["se"])
    exportDefaultCode = script.content.substring(imports[imports.length - 1]["se"])
  }
  // generate useCssVars code
  const useCssVarsCode = genCssVarsCode(attrs, id)
  const insertIndex = exportDefaultCode.indexOf("{") + 1
  // rewrite sfc file
  code = `<template>
${template.content}
</template>
<script>
${importCode}
${exportDefaultCode.slice(0, insertIndex)}${useCssVarsCode}${exportDefaultCode.slice(insertIndex)}
</script>
<style scoped ${Object.keys(styles[0].attrs).filter(key => key !== 'scoped')
                                        .map(key => `${key}='${styles[0].attrs[key]}'`)
                                        .join(" ")}>
${css}
</style>`

  return {
    code: code,
    map: null
  }
}
