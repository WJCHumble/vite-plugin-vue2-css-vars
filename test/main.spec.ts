import { transformMain } from "../src/main"

describe('transformMain', () => {
	const shortId = "data-v-7ba5bd90"
	const fileName = "App.vue"
	const baseSFC = `
<template>
  <div>
    <button @click="changeColor">change color</button>
  </div>
</template>

<script>
export default {
  data() {
    return {
      color: "blue",
    }
  },
  methods: {
    changeColor() {
      if (this.color === "red") {
        this.color = "black" 
      } else {
        this.color = "red"
      }
    }
  }
}
</script>
`.trim()

	test("should work when use vars attr in style", async () => {
		const sfc = `
${baseSFC}

<style scoped lang="less" vars>
.word {
	background: v-bind(color);

	p {
		background: v-bind(color2);
	}
}
</style>	
`
		const isProd = false
		const res = await transformMain(sfc, fileName, shortId, isProd)
		expect(res?.code).toMatchSnapshot()
	})

	test("In prod mode, will not generate cssVarsCode for css comment code when have vars attr in style without write v-bind() code", async () => {
		const sfc = `
${baseSFC}

<style scoped lang="less" vars>
.word {
	// background: v-bind(color);

	// p {
	// 	background: v-bind(color2);
	// }
}
</style>	
`
		const isProd = true
		const res = await transformMain(sfc, fileName, shortId, isProd)
		expect(res?.code).toMatchSnapshot()
	})

	test("In dev mode, will auto inject cssVarsCode when use vars attr in style without v-bind() code", async () => {
		const sfc = `
${baseSFC}

<style scoped lang="less" vars>
.word {
	// background: v-bind(color);

	// p {
	// 	background: v-bind(color2);
	// }
}
</style>	
`
		const isProd = false
		const res = await transformMain(sfc, fileName, shortId, isProd)
		expect(res?.code).toMatchSnapshot()
	})
})