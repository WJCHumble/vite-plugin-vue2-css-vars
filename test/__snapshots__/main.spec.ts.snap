// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformMain In dev mode, will auto inject cssVarsCode when use vars attr in style without v-bind() code 1`] = `
"<template>

<div>
  <button @click=\\"changeColor\\">change color</button>
</div>

</template>
<script>

//
//
//
//
//
//
//

export default {
  mixins: [{
    mounted: function () {
      if (!this._setVnodeVar) {
        this._setVnodeVar = function(vnode, cssVar, value) {
          if (!vnode.tag) {
            return
          }

          vnode.elm.style.setProperty(cssVar, value)

          if (vnode.children && vnode.children.length) {
            vnode.children.forEach((childVnode) => {
              this._setVnodeVar(childVnode, cssVar, value);
            });
          }
        }
      }
      
    }
  }],
  data() {
    return {
      color: \\"blue\\",
    }
  },
  methods: {
    changeColor() {
      if (this.color === \\"red\\") {
        this.color = \\"black\\" 
      } else {
        this.color = \\"red\\"
      }
    }
  }
}

</script>
<style scoped lang='less' vars='true'>



























.word {
	/* background: v-bind(color);*/

	/* p {*/
	/* 	background: v-bind(color2);*/
	/* }*/
}

</style>"
`;

exports[`transformMain In prod mode, will not generate cssVarsCode for css comment code when have vars attr in style without write v-bind() code 1`] = `undefined`;

exports[`transformMain should work when use vars attr in style 1`] = `
"<template>

<div>
  <button @click=\\"changeColor\\">change color</button>
</div>

</template>
<script>

//
//
//
//
//
//
//

export default {
  mixins: [{
    mounted: function () {
      if (!this._setVnodeVar) {
        this._setVnodeVar = function(vnode, cssVar, value) {
          if (!vnode.tag) {
            return
          }

          vnode.elm.style.setProperty(cssVar, value)

          if (vnode.children && vnode.children.length) {
            vnode.children.forEach((childVnode) => {
              this._setVnodeVar(childVnode, cssVar, value);
            });
          }
        }
      }
      
      this.$watch('color', function(nVal) { 
        const vnode = this._vnode 
        this._setVnodeVar(vnode, '--data-v-7ba5bd90-color', nVal)
      }, { immediate: true })
,
      this.$watch('color2', function(nVal) { 
        const vnode = this._vnode 
        this._setVnodeVar(vnode, '--data-v-7ba5bd90-color2', nVal)
      }, { immediate: true })

    }
  }],
  data() {
    return {
      color: \\"blue\\",
    }
  },
  methods: {
    changeColor() {
      if (this.color === \\"red\\") {
        this.color = \\"black\\" 
      } else {
        this.color = \\"red\\"
      }
    }
  }
}

</script>
<style scoped lang='less' vars='true'>



























.word {
	background: var(--data-v-7ba5bd90-color);

	p {
		background: var(--data-v-7ba5bd90-color2);
	}
}

</style>"
`;
